xarray/core/computation.py

```diff
@@ -0,0 +1,675 @@
+                      for arg in re.findall(ARGUMENT, arg_list)]
+                     for arg_list in string.split('->')])
+
+    def __eq__(self, other):
+        try:
+            return (self.input_core_dims == other.input_core_dims and
+                    self.output_core_dims == other.output_core_dims)
+        except AttributeError:
+            return False
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __repr__(self):
+        return ('%s(%r, %r)'
+                % (type(self).__name__,
+                   list(self.input_core_dims),
+                   list(self.output_core_dims)))
+
+
+def result_name(objects):
+    # type: List[object] -> Any
+    # use the same naming heuristics as pandas:
+    # https://github.com/blaze/blaze/issues/458#issuecomment-51936356
+    names = set(getattr(obj, 'name', None) for obj in objects)
+    names.discard(None)
+    if len(names) == 1:
+        name, = names
+    else:
+        name = None
+    return name
+
+
+_REPEAT_NONE = itertools.repeat(None)
+
+def build_output_coords(
+        args,                 # type: list
+        signature,            # type: Signature
+        new_coords=None,      # type: Iterable[Optional[Mapping]]
+        exclude=frozenset(),  # type: set
+        ):
+    # type: (...) -> List[OrderedDict[Any, Variable]]
+
+    if new_coords is None:
+        new_coords = _REPEAT_NONE
+
+    input_coords = []
+    for arg in args:
+        try:
+            coords = arg.coords
+        except AttributeError:
+            pass  # skip this argument
+        else:
+            coord_vars = getattr(coords, 'variables', coords)
+            input_coords.append(coord_vars)
+
+    output_coords = []
+    for output_dims, arg_new_coords in zip(
+            signature.output_core_dims, new_coords):
+
+        if len(input_coords) == 1 and not exclude and arg_new_coords is None:
+            # we can skip the expensive merge
+            unpacked_input_coords, = input_coords
+            merged = OrderedDict(unpacked_input_coords)
+        elif arg_new_coords:
+            merged = merge_coords_without_align([arg_new_coords] + input_coords,
+                                                priority_arg=0, exclude=exclude)
+        else:
+            merged = merge_coords_without_align(input_coords, exclude=exclude)
+
+        missing_dims = [dim for dim in output_dims if dim not in merged]
+        if missing_dims:
+            raise ValueError(
+                'new output dimensions must have matching entries '
+                'in `new_coords`: %r' % missing_dims)
+
+        dropped_dims = signature.all_input_core_dims - set(output_dims)
+        if dropped_dims:
+            merged = OrderedDict((k, v) for k, v in merged.items()
+                                 if dropped_dims.isdisjoint(v.dims))
+
+        output_coords.append(merged)
+
+    return output_coords
+
+
+def apply_dataarray_ufunc(func, *args, **kwargs):
```

Comment:
I've gone back and forth on this one.

The basic reason is that confusing to need to keep track of which functions can be called like `f(*args)` vs `f(args)`. Most functions we will be wrapping fall in the former category, at least if we want to be able to use real ufuncs (like `np.add`) directly. I'm using lots of `functools.partial` to pass around function objects, and I wanted to avoid thinking about this.

Note that it does cost us ~5 us to parse `**kwargs` manually. This is significant compared to the cost of some operations (e.g., a unary op on `Variable` with only a few elements takes only ~10 us otherwise) but becomes negligible once you start working doing operations that involve one or more `DataArray`/`Dataset` objects (in the 400 us range). I view this as the cost of writing Python -- none of it matters once you have big enough arrays.

Note once we switch to Python 3 only the overhead should go away and everything will look pretty clean, because we can drop `**kwargs`. Even before then, writing out the signature at the top of the docstring ensures that the API docs look right (there are also signature rewriting tricks we could use to get introspection/auto-complete working properly on Python 3, but that's probably overkill).